src/python/aqueduct_executor/operators/utils/execution.py

```diff
@@ -0,0 +1,131 @@
+import io
+import sys
+import traceback
+
+from contextlib import redirect_stderr, redirect_stdout
+from typing import Callable, Optional
+from pydantic import BaseModel
+from aqueduct_executor.operators.utils.enums import ExecutionStatus, FailureType
+
+
+_GITHUB_ISSUE_LINK = "https://github.com/aqueducthq/aqueduct/issues/new?assignees=&labels=bug&template=bug_report.md&title=%5BBUG%5D"
+
+TIP_OP_EXECUTION = "Error executing operator. Please refer to the stack trace for fix."
+_TIP_CREATE_BUG_REPORT = (
+    f"Please create bug report in github: {_GITHUB_ISSUE_LINK} . "
+    "We will get back to you as soon as we can."
+)
+TIP_UNKNOWN_ERROR = f"Sorry, we've run into an unexpected error! {_TIP_CREATE_BUG_REPORT}"
+TIP_INTEGRATION_CONNECTION = (
+    "We have trouble connecting to the integration. "
+    "Please check your credentials or contact your integraiton provider."
+)
+TIP_DEMO_CONNECTION = f"We have trouble connecting to demo DB. {_TIP_CREATE_BUG_REPORT}"
+
+TIP_EXTRACT = "We couldn't execute the provided query. Please double check your query is correct."
+TIP_LOAD = "We couldn't load to the integration. Please make sure the target exists, or you have the right permission."
+TIP_DISCOVER = "We couldn't list items in the integration. Please make sure your credentials have the right permission."
+
+
+class Error(BaseModel):
+    context: str = ""
+    tip: str = ""
+
+
+class Logs(BaseModel):
+    stdout: str = ""
+    stderr: str = ""
+
+
+class ExecutionState(BaseModel):
+    user_logs: Logs
+    status: ExecutionStatus = ExecutionStatus.PENDING
+    failure_type: FailureType = FailureType.SUCCESS
+    error: Optional[Error] = None
+
+    def user_fn_redirected(self, failure_tip: str) -> Callable:
+        def wrapper(user_fn: Callable) -> Callable:
+            def inner(*args, **kwargs):
+                stdout_log = io.StringIO()
+                stderr_log = io.StringIO()
+                try:
+                    with redirect_stdout(stdout_log), redirect_stderr(stderr_log):
+                        result = user_fn(*args, **kwargs)
+                except Exception:
+                    # Include the stack trace within the user's code.
+                    fetch_redirected_logs(stdout_log, stderr_log, self.user_logs)
+                    self.code = ExecutionStatus.FAILED
+                    self.failure_type = FailureType.USER
+                    self.error = Error(
+                        context=stack_traceback(
+                            offset=1
+                        ),  # traceback the first stack frame, which belongs to user
+                        tip=failure_tip,
+                    )
+                    print(f"User failure. Full log: {self.json()}")
+                    return None
+
+                # Include the stack trace within the user's code.
+                fetch_redirected_logs(stdout_log, stderr_log, self.user_logs)
```

Comment:
Do we need to call `fetch_redirected_logs() twice? Shouldn't a
```
finally:
    fetch_redirected_logs(...)
```
work the same way.

Reply:
Also `fetch_redirected_logs()` actually populates the logs argument, can we rename it `set_standard_logs(...)` or something?

Reply:
the subsequent steps of `fetch_redirected_logs()` are different when handling exceptions vs succeeded, so I don't think `finally` would work here.

Could you add more details on the second part? Don't quite get it :)

Reply:
Yeah I was just saying since the function takes in a stdout and stderr and writes those streams to `self.user_logs`, so its not really doing a fetch. A "set" might be more appropriate here.